File: src/test/java/com/koddy/server/auth/utils/TokenProviderTest.java
package com.koddy.server.auth.utils;

import com.koddy.server.auth.exception.AuthException;
import com.koddy.server.common.ParallelTest;
import com.koddy.server.member.domain.model.Member;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static com.koddy.server.auth.exception.AuthExceptionCode.INVALID_TOKEN;
import static com.koddy.server.common.fixture.MentorFixture.MENTOR_1;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;

@DisplayName("Auth -> TokenProvider 테스트")
class TokenProviderTest extends ParallelTest {
    private static final String SECRET_KEY = "asldfjsadlfjalksjf01jf02j9012f0120f12jf1j29v0saduf012ue101212c01";

    private final TokenProvider invalidProvider = new TokenProvider(SECRET_KEY, 0L, 0L);
    private final TokenProvider validProvider = new TokenProvider(SECRET_KEY, 7200L, 7200L);

    private final Member<?> member = MENTOR_1.toDomain().apply(1L);

    @Test
    @DisplayName("AccessToken과 RefreshToken을 발급한다")
    void createToken() {
        // when
        final String accessToken = validProvider.createAccessToken(member.getId(), member.getRoleTypes());
        final String refreshToken = validProvider.createRefreshToken(member.getId());

        // then
        assertAll(
                () -> assertThat(accessToken).isNotNull(),
                () -> assertThat(refreshToken).isNotNull()
        );
    }

    @Test
    @DisplayName("Token의 Payload를 추출한다")
    void extractPayload() {
        // when
        final String accessToken = validProvider.createAccessToken(member.getId(), member.getRoleTypes());

        // then
        assertThat(validProvider.getId(accessToken)).isEqualTo(member.getId());
    }

    @Test
    @DisplayName("Token 만료에 대한 유효성을 검증한다")
    void validateToken1() {
        // when
        final String validToken = validProvider.createAccessToken(member.getId(), member.getRoleTypes());
        final String invalidToken = invalidProvider.createAccessToken(member.getId(), member.getRoleTypes());

        // then
        assertDoesNotThrow(() -> validProvider.validateToken(validToken));
        assertThatThrownBy(() -> invalidProvider.validateToken(invalidToken))
                .isInstanceOf(AuthException.class)
                .hasMessage(INVALID_TOKEN.getMessage());
    }

    @Test
    @DisplayName("Token 조작에 대한 유효성을 검증한다")
    void validateToken2() {
        // when
        final String forgedToken = validProvider.createAccessToken(member.getId(), member.getRoleTypes()) + "hacked";

        // then
        assertThatThrownBy(() -> validProvider.validateToken(forgedToken))
                .isInstanceOf(AuthException.class)
                .hasMessage(INVALID_TOKEN.getMessage());
    }
}

File: src/test/java/com/koddy/server/auth/domain/service/TokenProviderTest.kt
package com.koddy.server.auth.domain.service

import com.koddy.server.auth.exception.AuthException
import com.koddy.server.auth.exception.AuthExceptionCode.INVALID_TOKEN
import com.koddy.server.common.UnitTestKt
import com.koddy.server.common.fixture.MentorFixtureStore.mentorFixture
import com.koddy.server.member.domain.model.Member
import io.kotest.assertions.assertSoftly
import io.kotest.assertions.throwables.shouldNotThrowAny
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.annotation.DisplayName
import io.kotest.core.spec.style.FeatureSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.throwable.shouldHaveMessage

@UnitTestKt
@DisplayName("Auth -> TokenProvider 테스트")
internal class TokenProviderTest : FeatureSpec({
    val secretKey = "asldfjsadlfjalksjf01jf02j9012f0120f12jf1j29v0saduf012ue101212c01"
    val invalidProvider = TokenProvider(
        secretKey = secretKey,
        accessTokenValidityInSeconds = 0L,
        refreshTokenValidityInSeconds = 0L,
    )
    val validProvider = TokenProvider(
        secretKey = secretKey,
        accessTokenValidityInSeconds = 7200L,
        refreshTokenValidityInSeconds = 7200L,
    )

    val member: Member<*> = mentorFixture(id = 1).toDomain()

    feature("TokenProvider's createAccessToken & createRefreshToken") {
        scenario("AccessToken과 RefreshToken을 발급한다") {
            val accessToken: String = validProvider.createAccessToken(member.id, member.authority)
            val refreshToken: String = validProvider.createRefreshToken(member.id)

            assertSoftly {
                accessToken shouldNotBe null
                refreshToken shouldNotBe null
            }
        }
    }

    feature("TokenProvider's getId & getAuthority") {
        scenario("토큰의 Payload(ID, Authority)를 추출한다") {
            val accessToken: String = validProvider.createAccessToken(member.id, member.authority)
            val refreshToken: String = validProvider.createRefreshToken(member.id)

            assertSoftly {
                validProvider.getId(accessToken) shouldBe member.id
                validProvider.getAuthority(accessToken) shouldBe member.authority
                validProvider.getId(refreshToken) shouldBe member.id
            }
        }
    }

    feature("TokenProvider's validateToken (Expired)") {
        scenario("토큰 만료에 대한 유효성 검사를 진행한다") {
            val liveToken: String = validProvider.createAccessToken(member.id, member.authority)
            val expiredToken: String = invalidProvider.createAccessToken(member.id, member.authority)

            assertSoftly {
                shouldNotThrowAny { validProvider.validateAccessToken(liveToken) }
                shouldThrow<AuthException> {
                    invalidProvider.validateAccessToken(expiredToken)
                } shouldHaveMessage INVALID_TOKEN.message
            }
        }
    }

    feature("TokenProvider's validateToken (Malformed)") {
        scenario("토큰 조작에 대한 유효성 검사를 진행한다") {
            val validToken: String = validProvider.createAccessToken(member.id, member.authority)
            val forgedToken: String = validProvider.createAccessToken(member.id, member.authority) + "hacked"

            assertSoftly {
                shouldNotThrowAny { validProvider.validateAccessToken(validToken) }
                shouldThrow<AuthException> {
                    validProvider.validateAccessToken(forgedToken)
                } shouldHaveMessage INVALID_TOKEN.message
            }
        }
    }

    feature("TokenProvider's validateToken (Different Subject)") {
        scenario("토큰 Subject에 대한 유효성 검사를 진행한다") {
            val accessToken: String = validProvider.createAccessToken(member.id, member.authority)
            val refreshToken: String = validProvider.createRefreshToken(member.id)

            assertSoftly {
                shouldNotThrowAny { validProvider.validateAccessToken(accessToken) }
                shouldNotThrowAny { validProvider.validateRefreshToken(refreshToken) }
                shouldThrow<AuthException> {
                    validProvider.validateAccessToken(refreshToken)
                } shouldHaveMessage INVALID_TOKEN.message
                shouldThrow<AuthException> {
                    validProvider.validateRefreshToken(accessToken)
                } shouldHaveMessage INVALID_TOKEN.message
            }
        }
    }
})
